### PR TITLE
improve error message for using property of type as type

### LINKED ISSUE #45333: `Namespace.Interface.prop` gives wrong error message
# Bug Report

### 🔎 Search Terms

namespace interface property dot

### 🕗 Version & Regression Information

- This is the behavior in every version I tried, and I reviewed the FAQ for entries about _________

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?ts=3.3.3#code/HYQwtgpgzgDiDGEAEA5c06KQbwFBIKQgA8YB7AJwBckBLYKiCgMwWQEkGnWs9D+kMCmRgAuJFCoV6Ac3yEAvriW4A9KqQBRCsIri0kWGyQByAxjYmkACxBQkwMkVKVGAEySQwAIyanOjCyWAHS4VACeMMgAQgCMSAC8qOhGiMEB3GzBQiIA3LhhkTEATInJhpgQ6VxBiADaJjkwJgC6+QW4JOTUdDU8HDjyBGJIIMDh+SrqWjqU4gDCY440CIhQ9ibs2Va+8CAArlDIm1a09iBIEVEANEje+zTLow4plcFIACK0HuFk+54QMaXJwUCBSWgQABuyCo1hhRSQZGYlzhgmEUWo4VMzV6-isAHdaLD-HUAEQwUktEwAfgAFIUooIylsYLkgA)

### 💻 Code

```ts
namespace Namespace {
    export interface Interface {
        prop: string
    }
}

// Error: Namespace 'Namespace' has no exported member 'Interface'.
type B1 = Namespace.Interface.prop;

type B2 = Namespace.Interface['prop'];



export interface I {
    p: any;
}

// Error: Cannot access 'I.p' because 'I' is a type, but not a namespace. Did you mean to retrieve the type of the property 'p' in 'I' with 'I["p"]'?(
type p = I.p;
```

### 🙁 Actual behavior

`type B1 = Namespace.Interface.prop;` gives `Namespace 'Namespace' has no exported member 'Interface'.`

### 🙂 Expected behavior

It should be `Cannot access 'Namespace.Interface.prop' because 'Namespace.Interface' is a type, but not a namespace. Did you mean to retrieve the type of the property 'prop' in 'Namespace.Interface' with 'Namespace.Interface["prop"]'?`

### PR DESCRIPTION
**Pull request by Zzzen**
Fixes #45333

### CODE REVIEW

File: tests/baselines/reference/errorForUsingPropertyOfTypeAsType03.errors.txt [hunk 1]
```diff
@@ -19,8 +19,8 @@ tests/cases/compiler/errorForUsingPropertyOfTypeAsType03.ts(24,13): error TS2713
         type C2 = typeof Color;
     
         let a1: Color.Red.toString;
-                      ~~~
-!!! error TS2694: Namespace 'Color' has no exported member 'Red'.
+                          ~~~~~~~~
+!!! error TS2713: Cannot access 'Red.toString' because 'Red' is a type, but not a namespace. Did you mean to retrieve the type of the property 'toString' in 'Red' with 'Red["toString"]'?
```
Comment (DanielRosenwasser) [inline tests/baselines/reference/errorForUsingPropertyOfTypeAsType03.errors.txt:23]: I think we'd really like to suggest `typeof Color.Red.toString` first if that would be valid, then fall back to using indexed access syntax.

**Review by andrewbranch — APPROVED** [review]
Wow, what an improvement 🌟

File: src/compiler/checker.ts [hunk 2]
```diff
@@ -3148,6 +3148,30 @@ namespace ts {
             return symbol.parent ? getFullyQualifiedName(symbol.parent, containingLocation) + "." + symbolToString(symbol) : symbolToString(symbol, containingLocation, /*meaning*/ undefined, SymbolFormatFlags.DoNotIncludeSymbolChain | SymbolFormatFlags.AllowAnyNodeKind);
         }
 
+        function getFullQualifiedName(node: QualifiedName) {
+            while(isQualifiedName(node.parent)) {
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts:3152]:
```suggestion
            while (isQualifiedName(node.parent)) {
```

Comment (DanielRosenwasser) [inline src/compiler/checker.ts:3152]:
```suggestion
            while (isQualifiedName(node.parent)) {
```

File: src/compiler/checker.ts [hunk 3]
```diff
@@ -3148,6 +3148,30 @@ namespace ts {
             return symbol.parent ? getFullyQualifiedName(symbol.parent, containingLocation) + "." + symbolToString(symbol) : symbolToString(symbol, containingLocation, /*meaning*/ undefined, SymbolFormatFlags.DoNotIncludeSymbolChain | SymbolFormatFlags.AllowAnyNodeKind);
         }
 
+        function getFullQualifiedName(node: QualifiedName) {
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts:3151]:
```suggestion
        function getContainingQualifiedNameNode(node: QualifiedName) {
```

File: src/compiler/checker.ts [hunk 4]
```diff
@@ -3199,9 +3223,14 @@ namespace ts {
                         const declarationName = declarationNameToString(right);
                         const suggestionForNonexistentModule = getSuggestedSymbolForNonexistentModule(right, namespace);
                         const exportedTypeSymbol = getMergedSymbol(getSymbol(getExportsOfSymbol(namespace), right.escapedText, SymbolFlags.Type));
+                        const fullQualifiedName = isQualifiedName(name) && getFullQualifiedName(name);
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts:3226]:
```suggestion
                        const fullQualifiedName = isQualifiedName(name) && getContainingQualifiedNameNode(name);
```

**Review by DanielRosenwasser — CHANGES_REQUESTED** [review]
Looks great except for minor changes I'd like to see before merging.

File: src/compiler/checker.ts [hunk 5]
```diff
@@ -3197,10 +3221,22 @@ namespace ts {
...
+                        const suggestionForNonexistentModule = getSuggestedSymbolForNonexistentModule(right, namespace);
+                        const exportedTypeSymbol = getMergedSymbol(getSymbol(getExportsOfSymbol(namespace), right.escapedText, SymbolFlags.Type));
+                        const fullQualifiedName = isQualifiedName(name) && getFullQualifiedName(name);
+                        const canSuggestTypeof = fullQualifiedName && !isTypeOfExpression(fullQualifiedName.parent) && tryGetQualifiedNameAsValue(fullQualifiedName);
+                        if (suggestionForNonexistentModule) {
+                            error(right, Diagnostics._0_has_no_exported_member_named_1_Did_you_mean_2, namespaceName, declarationName, symbolToString(suggestionForNonexistentModule));
+                        }
+                        else if (canSuggestTypeof) {
+                            error(name, Diagnostics._0_refers_to_a_value_but_is_being_used_as_a_type_here_Did_you_mean_typeof_0, entityNameToString(fullQualifiedName as QualifiedName));
```
Comment (Zzzen) [inline src/compiler/checker.ts:3232]: Haven't tried TypeScript 4.4, so I wrote `fullQualifiedName as QualifiedName`. After upgrading vscode, I find it is unnecessary. The CFA is really impressive, kudos to you guys!

**Review by DanielRosenwasser — APPROVED** [review]
What an improvement - thank you!

**Pull request merged**